src/style-spec/expression/evaluation_context.js

```diff
@@ -52,10 +54,11 @@ class EvaluationContext {
     }
 
     distanceFromCamera() {
```

Comment:
Is `distanceFromCamera` a descriptive name? It seems like this is used to compute the distance from the view center/look at rather than the distance from the camera. 

What catches my attention is that the difference seems to add a strong radial dependence to the cutoff. Is this the main criteria people would use to clip based on distance? I'm not opposed to it and can imagine that some of the factors which made fog challenging may factor into the decision, but I'm at least weighing the mental overhead of another similar but slightly differently defined distance field. Are there reasons not to use a distance metric similar in definition to fog?

![less-than-0 5](https://user-images.githubusercontent.com/572717/135532948-0f69559b-d3a5-4e09-bcaa-423cdcb07af2.png)

Reply:
Because this is how the fog distance field shifts around with pitch (this isnt a problem for fog because we only use it at high pitch where it doesnt shift) The shift is because of the translation the camera when pitching.
At pitch 0:
<img width="674" alt="Screen Shot 2021-09-30 at 3 48 31 PM" src="https://user-images.githubusercontent.com/1449257/135540111-c82abad0-f169-4359-b3a3-7c4c2a9641e5.png">
At pitch 20:
<img width="675" alt="Screen Shot 2021-09-30 at 3 48 48 PM" src="https://user-images.githubusercontent.com/1449257/135540117-a8c9c44c-5515-41ca-910d-307930c372e3.png">
At pitch 60:
<img width="674" alt="Screen Shot 2021-09-30 at 3 49 10 PM" src="https://user-images.githubusercontent.com/1449257/135540131-efa6ca43-cf48-4b74-aa46-4ae20518516c.png">
At pitch 80:
<img width="674" alt="Screen Shot 2021-09-30 at 3 49 24 PM" src="https://user-images.githubusercontent.com/1449257/135540142-18b03279-62c8-4d66-bcad-c58a025f0972.png">


Whereas this stays locked at the center.

Reply:
This is what I used to generate the geojson for the above viz
```
function generateDistanceScales() {
    const center = map.getCenter();
    const bearing = map.getBearing();
    const rings = 20;
    const step = 0.1;

    const matrix = glMatrix.mat4.invert([], map.transform.mercatorFogMatrix);
    const circles = [];
    for (let i = 1; i <= rings; i++) {
        const radius = step * i;

        for(let th = 0; th < 360; th+= 20) {
            const p = [radius * Math.sin(th * Math.PI/180), radius * Math.cos(th * Math.PI/180), 0];
            const coord = new mapboxgl.MercatorCoordinate(...glMatrix.vec3.transformMat4([], p, matrix)).toLngLat();
            const f = turf.point([coord.lng, coord.lat]);
            f.properties['distance'] =  `${(radius - 0.5 / Math.tan(map.transform.fov * 0.5)).toFixed(2)}`;
            circles.push(f);
        }
    }

    return turf.featureCollection(circles);
}
```

Reply:
@arindam1993 What's the advantage of this staying locked at the center? The fog behavior seems closer to a distance from camera operation. Wouldn't that be more intuitive for users than the radial behavior from map center, particularly if it could be consistent with fog?

Reply:
Because unlike fog this isn't a `high-pitch only` use-case, users can do whatever they want with the `pitch` and set much lower thresholds too if they wish. So imo having a nice and fixed reference frame is better so that the clipping boundaries arent moving around as much when pitching in the range of (0-40) deg.

Reply:
If you look at the screenshot above the value of the distance at the same location is changing around a lot.

Reply:
What's the use case of this without high pitch?

Reply:
Its still usable at `medium` pitch (30-60 deg), where  fog doesnt render fully but symbol clipping is useful still
Example:
30 deg pitch
<img width="675" alt="Screen Shot 2021-09-30 at 5 38 28 PM" src="https://user-images.githubusercontent.com/1449257/135548679-bd08672d-5242-4a80-9452-1a07286d4c8f.png">
40 deg pitch
<img width="674" alt="Screen Shot 2021-09-30 at 5 38 41 PM" src="https://user-images.githubusercontent.com/1449257/135548685-e2a68f83-735f-46fb-a371-1ebb5ad3f8e1.png">
60 deg pitch
<img width="674" alt="Screen Shot 2021-09-30 at 5 38 49 PM" src="https://user-images.githubusercontent.com/1449257/135548697-1f011405-5774-43d6-824c-9d77b4b6faa6.png">
and if you take the tiny patch of green as the reference point you can see the distance value there changing a lot

and if you're trying to clip symbols there this shifting would make it pretty weird

Reply:
I've created a notebook here which illustrates fog units: https://observablehq.com/d/50ec70cf44480b1d

The behavior at medium and high pitch seems preferable to me. The units make a little less sense at very low pitch, though the symbol-clip feature also makes a lot less sense to me at zero pitch as well. It would be possible to transition to viewport-fixed units (viewport size = 1.0) at zero pitch so that they would always make sense, but again, I'd want to be certain of the use case of symbol-clip at zero pitch before strongly suggesting that.

![fog-pitch-units](https://user-images.githubusercontent.com/572717/135651498-284ed3c7-9688-4731-8424-b801035985fd.gif)

Reply:
Talked to Arindam. One downside of the above is that it adds additional dependence on pitch, which would be nice to avoid. We're currently trying out ground-fixed planar lines (no pitch-dependence).

Reply:
I'm still struggling to understand the difference between how these approaches will effect feature filtering. Do we have an example that more closely reflects real-world use case of this feature?

Reply:
Got the planar lines implemented and tests updated.

Reply:
@SnailBones the problem is the foreshortening  of the far distances.
Sample case:
filter: `[ '<', ['distance-from-center'], 0.5]`
Here the value of 0.5 means that: `Clip this symbol in the region beyond what  would be the top  of the screen at 0 pitch`

But since the fog distance measures a 3d distance from the camera to the plane, the value of the distance at the same geographic location increases as pitch increases, which means that what wasn't being clipped at 0 pitch gets clipped at a higher pitch. You can see this in the GIF @rreusser posted, the first line corresponds to fog_distance=0.5, it starts of at the water at 0 pitch and then gets squished inwards.